cmd/boulder-wfe2/main_test.go

```diff
@@ -2,14 +2,85 @@ package main
 
 import (
 	"bytes"
+	"crypto/x509"
+	"encoding/pem"
 	"fmt"
 	"io/ioutil"
 	"strings"
 	"testing"
 
+	"github.com/letsencrypt/boulder/core"
+	"github.com/letsencrypt/boulder/issuance"
 	"github.com/letsencrypt/boulder/test"
 )
 
+func TestLoadChain_Valid(t *testing.T) {
+	issuer, chainPEM, err := loadChain([]string{
+		"../../test/test-ca-cross.pem",
+		"../../test/test-root2.pem",
+	})
+	test.AssertNotError(t, err, "Should load valid chain")
+
+	expectedIssuer, err := core.LoadCert("../../test/test-ca-cross.pem")
+	test.AssertNotError(t, err, "Failed to load test issuer")
+
+	chainIssuerPEM, rest := pem.Decode(chainPEM)
+	test.AssertNotNil(t, chainIssuerPEM, "Failed to decode chain PEM")
+	parsedIssuer, err := x509.ParseCertificate(chainIssuerPEM.Bytes)
+	test.AssertNotError(t, err, "Failed to parse chain PEM")
+
+	// The three versions of the intermediate (the one loaded by us, the one
+	// returned by loadChain, and the one parsed from the chain) should be equal.
+	test.AssertByteEquals(t, issuer.Raw, expectedIssuer.Raw)
+	test.AssertByteEquals(t, parsedIssuer.Raw, expectedIssuer.Raw)
+
+	// The chain should contain nothing else.
+	rootIssuerPEM, rest := pem.Decode(rest)
+	var nilBlock *pem.Block = nil
```

Comment:
Is there a better way to do this? I can't just put `nil` in the line below, as go complains about "comparing different types `*pem.Block` and `nil`".

Reply:
To be pedantic for a sec: It's not Go that complains, it's our `AssertEquals` helper:

```go
// AssertEquals uses the equality operator (==) to measure one and two
func AssertEquals(t *testing.T, one interface{}, two interface{}) {
	t.Helper()
	if reflect.TypeOf(one) != reflect.TypeOf(two) {
		t.Fatalf("cannot test equality of different types: %T != %T", one, two)
	}
	if one != two {
		t.Fatalf("%#v != %#v", one, two)
	}
}
```

This is the "[boxed nil](https://www.youtube.com/watch?v=ynoY2xz-F8s)" surprising result again. If you call `test.AssertEquals(..., nil)`, the parameter `two` is bound to an interface value equal to nil. In other words, nothing there: no dynamic type, no dynamic contents.

If you call `test.AssertEquals(..., nilBlock)`, the parameter `two` is bound to a non-nil interface value. That interface value has a dynamic type of `*pem.Block`, and a dynamic value of `nil`.

However, if you get rid of the `AssertEquals` call and just do the basic Go thing, it will work:

```
if rootIsserPEM != nil {
  t.Error("expected rootIssuerPEM to be nil but it was not")
}
```

That's because this code doesn't try to box up `nil` into an `interface{}` by calling `AssertEquals`.

https://golang.org/ref/spec#Type_assertions

Reply:
Ah of course. I knew it was "one of these is nil and the other is an interface whose value is nil", but I couldn't figure out why that was the case, since `pem.Block` is just a struct, not an interface. But of course passing it into a function that takes `interface{}` as its parameter does that for you. Thanks, updated.